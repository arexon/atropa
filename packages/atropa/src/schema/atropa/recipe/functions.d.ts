import type {
	RecipeBrewingContainer,
	RecipeBrewingMix,
	RecipeMaterialReduction,
	RecipeShaped,
	RecipeShapeless,
	RecipeFurnace
} from '../../../vanilla/recipe'

type FurnaceFunction = {
	/**
	 * ## Recipe Furnace
	 * Represents a recipe for use with a Furnace.
	 */
	furnace: (template: RecipeFurnace) => void
}

type ShapedFunction = {
	/**
	 * ## Recipe Shaped
	 * Represents a recipe that requires a dedicated pattern when using a Crafting Table.
	 */
	shaped: (template: RecipeShaped) => void
}

type ShapelessFunction = {
	/**
	 * ## Recipe Shapeless
	 * Represents a recipe that does not require a dedicated pattern.
	 */
	shapeless: (template: RecipeShapeless) => void
}

type BrewingMixFunction = {
	/**
	 * ## Recipe Brewing Mix
	 * Represents a recipe that for use with a Potion Brewing station.
	 */
	brewingMix: (template: RecipeBrewingMix) => void
}

type BrewingContainerFunction = {
	/**
	 * ## Recipe Brewing Container
	 * Represents a recipe that for use with a Potion Brewing station.
	 */
	brewingContainer: (template: RecipeBrewingContainer) => void
}

type MaterialReductionFunction = {
	/**
	 * ## Recipe Material Reduction
	 */
	materialReduction: (template: RecipeMaterialReduction) => void
}

export type {
	FurnaceFunction,
	ShapedFunction,
	ShapelessFunction,
	BrewingMixFunction,
	BrewingContainerFunction,
	MaterialReductionFunction
}
