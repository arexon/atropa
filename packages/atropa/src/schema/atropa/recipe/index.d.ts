export type { RecipeTemplate } from './template'
