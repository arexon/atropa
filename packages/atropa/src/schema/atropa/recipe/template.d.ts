import type { Namespace } from '../../common/template'
import type {
	FurnaceFunction,
	ShapedFunction,
	ShapelessFunction,
	BrewingMixFunction,
	BrewingContainerFunction,
	MaterialReductionFunction
} from './functions'

interface RecipeTemplate
	extends Namespace,
		FurnaceFunction,
		ShapedFunction,
		ShapelessFunction,
		BrewingMixFunction,
		BrewingContainerFunction,
		MaterialReductionFunction {}

export type { RecipeTemplate }
