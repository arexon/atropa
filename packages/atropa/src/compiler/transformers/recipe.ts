import { objectMap } from '@antfu/utils'

interface UserTemplate {
	namespace: string
	shaped: (template: Record<string, any>) => void
	shapeless: (template: Record<string, any>) => void
	furnace: (template: Record<string, any>) => void
	brewingContainer: (template: Record<string, any>) => void
	brewingMix: (template: Record<string, any>) => void
	materialReduction: (template: Record<string, any>) => void
}

interface VanillaTemplate {
	recipe?: Record<string, any>
}
interface RecipeObject {
	format_version: string
	'minecraft:recipe_shaped'?: VanillaTemplate
	'minecraft:recipe_shapeless'?: VanillaTemplate
	'minecraft:recipe_furnace'?: VanillaTemplate
	'minecraft:recipe_brewing_container'?: VanillaTemplate
	'minecraft:recipe_brewing_mix'?: VanillaTemplate
	'minecraft:recipe_material_reduction'?: VanillaTemplate
}

export function processTemplate(template: VanillaTemplate): UserTemplate {
	return {
		namespace: process._namespace,
		shaped: (_template) => {
			template.recipe = { shaped: _template }
		},
		shapeless: (_template) => {
			template.recipe = { shapeless: _template }
		},
		furnace: (_template) => {
			template.recipe = { furnace: _template }
		},
		brewingContainer: (_template) => {
			template.recipe = { brewing_container: _template }
		},
		brewingMix: (_template) => {
			template.recipe = { brewing_mix: _template }
		},
		materialReduction: (_template) => {
			template.recipe = { material_reduction: _template }
		}
	}
}

export function transformTemplate(
	template: VanillaTemplate,
	version: string
): RecipeObject {
	const transformedTemplate = objectMap(
		template.recipe as Required<VanillaTemplate>,
		(key, value) => {
			return [`minecraft:recipe_${key}`, value]
		}
	)

	return {
		format_version: version,
		...transformedTemplate
	}
}
