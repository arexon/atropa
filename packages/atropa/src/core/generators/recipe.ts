import { objectMap } from '@antfu/utils'
import type { RecipeTemplate } from '../../schema/atropa/recipe'

interface UserTemplate {
	namespace: string
	shaped: (template: Record<string, any>) => void
	shapeless: (template: Record<string, any>) => void
	furnace: (template: Record<string, any>) => void
	brewingContainer: (template: Record<string, any>) => void
	brewingMix: (template: Record<string, any>) => void
	materialReduction: (template: Record<string, any>) => void
}
interface VanillaTemplate {
	recipe?: Record<string, any>
}
interface Recipe {
	format_version: string
	'minecraft:recipe_shaped'?: VanillaTemplate
	'minecraft:recipe_shapeless'?: VanillaTemplate
	'minecraft:recipe_furnace'?: VanillaTemplate
	'minecraft:recipe_brewing_container'?: VanillaTemplate
	'minecraft:recipe_brewing_mix'?: VanillaTemplate
	'minecraft:recipe_material_reduction'?: VanillaTemplate
}

/**
 * # Define Recipe
 *
 * Generates a recipe from the given template.
 * @param fn A callback function with function parameters used to define the recipe.
 * @returns A recipe.
 */
export function defineRecipe(
	fn: (template: RecipeTemplate) => void
): Record<string, any> {
	try {
		const template = {}

		fn(processTemplate(template))
		return transformTemplate(template, '1.12.0')
	} catch (error) {
		throw new Error(`Failed to transform recipe:`, error as Error)
	}
}

function processTemplate(template: VanillaTemplate): UserTemplate {
	return {
		namespace: atropa.config.namespace,
		shaped: (_template) => {
			template.recipe = { shaped: _template }
		},
		shapeless: (_template) => {
			template.recipe = { shapeless: _template }
		},
		furnace: (_template) => {
			template.recipe = { furnace: _template }
		},
		brewingContainer: (_template) => {
			template.recipe = { brewing_container: _template }
		},
		brewingMix: (_template) => {
			template.recipe = { brewing_mix: _template }
		},
		materialReduction: (_template) => {
			template.recipe = { material_reduction: _template }
		}
	}
}

function transformTemplate(template: VanillaTemplate, version: string): Recipe {
	const transformedTemplate = objectMap(
		template.recipe as Required<VanillaTemplate>,
		(key, value) => {
			return [`minecraft:recipe_${key}`, value]
		}
	)

	return {
		format_version: version,
		...transformedTemplate
	}
}
