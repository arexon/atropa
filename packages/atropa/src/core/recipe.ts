import {
	processTemplate,
	transformTemplate
} from '../compiler/transformers/recipe'
import type { RecipeTemplate } from '../schema/atropa/server/recipe'

/**
 * # Define Recipe
 *
 * Generates a recipe from the given template.
 * @param fn A callback function with function parameters used to define the recipe.
 * @returns A recipe.
 */
export function defineRecipe(
	fn: (template: RecipeTemplate) => void
): Record<string, any> {
	try {
		const template = {}

		fn(processTemplate(template))
		return transformTemplate(template, '1.12.0')
	} catch (error) {
		throw new Error(`Failed to transform recipe:`, error as Error)
	}
}
