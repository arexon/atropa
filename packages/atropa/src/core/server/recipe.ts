import { logger } from '../../logger'
import type { RecipeTemplate } from '../../schema/atropa/server/recipe'

interface VanillaTemplate {
	recipe?: Record<string, any>
}

/**
 * # Define Recipe
 *
 * Generates a recipe from the given template.
 * @param fn A callback function with function parameters used to define the recipe.
 * @returns A recipe.
 */
export function defineRecipe(
	fn: (template: RecipeTemplate) => void
): Record<string, any> {
	try {
		const template: VanillaTemplate = {}

		fn(processTemplate(template))

		return { format_version: '1.12.0', ...template.recipe }
	} catch (error) {
		logger.error(`Failed to parse recipe:`, error)
		process.exit(1)
	}
}

function processTemplate(fields: VanillaTemplate): RecipeTemplate {
	const getObject = (name: string, template: Record<string, any>) => {
		return { [`minecraft:recipe_${name}`]: template }
	}
	return {
		namespace: global.config.namespace,
		shaped: (template) => {
			fields.recipe = getObject('shaped', template)
		},
		shapeless: (template) => {
			fields.recipe = getObject('shapeless', template)
		},
		furnace: (template) => {
			fields.recipe = getObject('furnace', template)
		},
		brewingContainer: (template) => {
			fields.recipe = getObject('brewing_container', template)
		},
		brewingMix: (template) => {
			fields.recipe = getObject('brewing_mix', template)
		},
		materialReduction: (template) => {
			fields.recipe = getObject('material_reduction', template)
		}
	}
}
