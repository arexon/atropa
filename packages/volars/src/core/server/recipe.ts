import { logger } from '../../logger'
import type { RecipeTemplate } from '../../schema/volars/server/recipe/template'

export function defineRecipe(
	fn: (template: RecipeTemplate) => void
): Record<string, any> {
	let template: Record<string, any> = {}

	try {
		fn({
			namespace: global.config.namespace,
			shaped: (_template) => {
				template = { [`minecraft:recipe_shaped`]: _template }
			},
			shapeless: (_template) => {
				template = { [`minecraft:recipe_shapeless`]: _template }
			},
			furnace: (_template) => {
				template = { [`minecraft:recipe_furnace`]: _template }
			},
			brewingContainer: (_template) => {
				template = { [`minecraft:recipe_brewing_container`]: _template }
			},
			brewingMix: (_template) => {
				template = { [`minecraft:recipe_brewing_mix`]: _template }
			},
			materialReduction: (_template) => {
				template = {
					[`minecraft:recipe_material_reduction`]: _template
				}
			}
		})

		return { format_version: '1.12', ...template }
	} catch (error) {
		logger.error(`Failed to parse recipe:`, error)
		process.exit(0)
	}
}
