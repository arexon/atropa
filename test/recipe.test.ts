import { describe, expect, it } from 'vitest'
import recipe from '#playground/BP/recipes/radio.recipe'

describe('define recipe', () => {
	it('returns a valid recipe object', () => {
		expect(recipe).toMatchSnapshot()
	})
})
