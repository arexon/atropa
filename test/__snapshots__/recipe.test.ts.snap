// Vitest Snapshot v1

exports[`define recipe > returns a valid recipe object 1`] = `
{
  "format_version": "1.12",
  "minecraft:recipe_shaped": {
    "description": {
      "identifier": "test:recipe.radio",
    },
    "key": {
      "A": {
        "item": "minecraft:planks",
      },
      "B": {
        "item": "minecraft:note",
      },
    },
    "pattern": [
      "A",
      "B",
    ],
    "result": {
      "item": "test:radio",
    },
    "tags": [
      "crafting_table",
    ],
  },
}
`;
